src/server.rs

```diff
@@ -0,0 +1,230 @@
+
+                        // TODO: invalid pixel formats?
+                        output_pixel_format = pf;
+                    }
+                    SetEncodings(e) => {
+                        debug!("Rx [{:?}]: SetEncodings={:?}", addr, e);
+                    }
+                    FramebufferUpdateRequest(f) => {
+                        debug!("Rx [{:?}]: FramebufferUpdateRequest={:?}", addr, f);
+
+                        let mut fbu = self.server.get_framebuffer_update().await;
+                        let data = self.data.lock().await;
+
+                        // We only need to change pixel formats if the client requested a different
+                        // one than what's specified in the input.
+                        //
+                        // For now, we only support transformations between 4-byte RGB formats, so
+                        // if the requested format isn't one of those, we'll just leave the pixels
+                        // as is.
+                        if data.input_pixel_format != output_pixel_format
+                            && data.input_pixel_format.is_rgb_888()
+                            && output_pixel_format.is_rgb_888()
+                        {
+                            debug!(
+                                "transforming: input={:#?}, output={:#?}",
+                                data.input_pixel_format, output_pixel_format
+                            );
+                            fbu = fbu.transform(&data.input_pixel_format, &output_pixel_format);
+                        }
+
+                        if let Err(e) = fbu.write_to(s).await {
+                            error!(
+                                "[{:?}] could not write FramebufferUpdateRequest: {:?}",
+                                addr, e
+                            );
+                            return;
+                        }
+                        debug!("Tx [{:?}]: FramebufferUpdate", addr);
+                    }
+                    KeyEvent(ke) => {
+                        trace!("Rx [{:?}]: KeyEvent={:?}", addr, ke);
+                    }
+                    PointerEvent(pe) => {
+                        trace!("Rx [{:?}: PointerEvent={:?}", addr, pe);
+                    }
+                    ClientCutText(t) => {
+                        trace!("Rx [{:?}: ClientCutText={:?}", addr, t);
+                    }
+                },
+                Err(e) => {
+                    error!("[{:?}] error reading client message: {}", addr, e);
+                    return;
+                }
+            }
+        }
+    }
+
+    pub async fn start(&self) {
+        let listener = TcpListener::bind(self.config.addr).await.unwrap();
+
+        loop {
+            let (mut s, a) = listener.accept().await.unwrap();
+            let server = self.clone();
+            tokio::spawn(async move {
+                VncServer::handle_conn(&server, &mut s, a).await;
```

Comment:
Nit: don't need to pass in the address here since anyone with the `TcpStream` can just call `peer_addr` to get the address if needed.

Reply:
hm, thanks for that suggestion. It looks like that method returns a result, so it would have to be handled, and i use the address in a bunch of log messages so we can track which client a message is from. is it cool if i leave it like this for now? I find it a bit cleaner.